File: python/src/graphs/breadth_first_search.py
class Node:
    def __init__(self, name) -> None:
        self.children = []
        self.name = name

    def add_child(self, name):
        self.children.append(Node(name))

    def breadth_first_search(self, array):
        queue = [self]

        while len(queue) > 0:
            current = queue.pop(0)
            array.append(current.name)
            for child in current.children:
                queue.append(child)
        return array


# Test
def test_breadth_first_search():
    # Create the root node and its children
    root = Node("A")
    root.add_child("B")
    root.add_child("C")
    root.children[0].add_child("D")
    root.children[0].add_child("E")
    root.children[1].add_child("F")

    # Expected order of nodes after a breadth-first search
    expected_order = ["A", "B", "C", "D", "E", "F"]

    # Perform the breadth-first search
    result = root.breadth_first_search([])

    # Assert that the result matches the expected order
    assert (
        result == expected_order
    ), f"Expected order {expected_order}, but got {result}"

File: rust/src/graphs/breadth_first_search.rs
use std::collections::VecDeque;

pub struct Node {
    name: String,
    children: Vec<Box<Node>>,
}

impl Node {
    pub fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            children: Vec::new(),
        }
    }

    pub fn add_child(&mut self, name: &str) -> &mut Self {
        self.children.push(Box::new(Node::new(name)));
        self
    }

    pub fn breadth_first_search(&self) -> Vec<String> {
        let mut queue = VecDeque::new();
        let mut array = Vec::new();

        queue.push_back(self);

        while let Some(current) = queue.pop_front() {
            array.push(current.name.clone());

            for child in &current.children {
                queue.push_back(child);
            }
        }

        array
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn ex1() {
        let mut root = Node::new("A");
        root.add_child("B").add_child("C");
        root.children[0].add_child("D");
        root.children[0].add_child("E");
        root.children[1].add_child("F");

        let expected_order =
            vec!["A", "B", "C", "D", "E", "F"];

        let result = root.breadth_first_search();

        assert_eq!(result, expected_order);
    }
}
